### ISSUE TITLE
Occasional  STF execution failure: InvalidNonce(2)

### ISSUE THREAD

**Issue by Kailai-Wang** [litentry/litentry-parachain#1175]
### Context

Occasionally CI failed because some STFs weren't executed successfully, e.g. [this run](https://github.com/litentry/litentry-parachain/actions/runs/3881542642/attempts/1) - I re-attach the logs.

The demo log shows that after the transfer the balance is unchanged: 

```
Get balance of Alice's incognito account
9000000000000

Get balance of Bob's incognito account
0

* Send 2000000000000 funds from Alice's incognito account to Bob's incognito account
send trusted call transfer from 5Dt1Wg85pXGLstt36t6TDdvXXoCtG6zkUL17KkyVaPYSrzGH to 5CtBzV6DZQUqhtiq3XAMPgYkcAfrbSuxPjjz4PAs98tuzSVB: 2000000000000

* Get balance of Alice's incognito account
9000000000000
```

In worker log there's something like
```
[2023-01-10T10:50:48Z ERROR itp_stf_executor::executor] Stf execute failed: InvalidNonce(2)
```

Questions are:
- why did the transfer fail?
- is it a timing problem? maybe waiting for 30sec is not enough? But it can't explain why it happened upstream too: https://github.com/integritee-network/worker/actions/runs/3882003200/jobs/6622681953, where the block interval is 6s
- was the `InvalidNonce` error the cause or result of the previous failure? I could imagine that if an STF fails to execute, the system nonce won't be incremented, but the nonce is the trusted-call is already pre-calculated when signing => hence the mismatch

---
:heavy_check_mark: Please set appropriate **labels** and **assignees** if applicable.
[logs-M6-sidechain.zip](https://github.com/litentry/litentry-parachain/files/10384401/logs-M6-sidechain.zip)

**Comment by Kailai-Wang** [comment]
Who wants to investigate it? @litentry/tee 

It **can** be a difficult one

Maybe related: https://github.com/integritee-network/worker/issues/1132
but it's sidechain test upstream

**Comment by zTgx** [comment]
CI erros:

---

Failed in first case:
https://github.com/integritee-network/worker/actions/runs/3882003200/jobs/6622681953

Init Balance: 
A=50000000000
B=0

A->send->B 25000000000

CheckBalance:
A=50000000000
B=0

A->unshield 15000000000
sleep 10s

CheckBalance:
A=25000000000(should be 10000000000)
B=?(no output)


----


Failed in second case:
https://github.com/litentry/litentry-parachain/actions/runs/3903235035/jobs/6669901892

After run firstcase succesfully:

Balance:
A=9000000000000
B=0

A->send->B 2000000000000

CheckBalance:
A=9000000000000
B=0

A->unshield 1000000000000
sleep 30s

CheckBalance:
A=7000000000000(should be 60000000000)
B=2000000000000

**Comment by zTgx** [comment]
~~Get Incorrect balance of incognito accounts at my local test:~~

need to wait for at least one block interval.

---

Using client binary ./../bin/integritee-cli
Using node uri ws://127.0.0.1:9944
Using trusted-worker uri wss://127.0.0.1:2000

* Query on-chain enclave registry:
Reading MRENCLAVE from worker list: 47M7fLLfdKVXYY2yB7SMZsYmqdvNcWSgBY871QUe48jS
* Create a new incognito account for Alice
  Alice's incognito account = //AliceIncognito

* Create a new incognito account for Bob
  Bob's incognito account = 5FX2Bk6yjErUQ2H9iMH1t1nJnRdY75kjgD5D82S6NFhicCTr


* Shield 8000000000000 tokens to Alice's incognito account
[+] TrustedOperation got finalized. Hash: Some(0xdf8f571d0715879872f4649136dd63785f9abbfebb140466f6068f7dd39d6b2b)


**Get balance of Alice's incognito account**
**0**

**Get balance of Bob's incognito account**
**0**

* Send 2000000000000 funds from Alice's incognito account to Bob's incognito account
send trusted call transfer from 5Dt1Wg85pXGLstt36t6TDdvXXoCtG6zkUL17KkyVaPYSrzGH to 5FX2Bk6yjErUQ2H9iMH1t1nJnRdY75kjgD5D82S6NFhicCTr: 2000000000000

* Get balance of Alice's incognito account
6000000000000

* Bob's incognito account balance
2000000000000

**Comment by Kailai-Wang** [comment]
I think I get it, the problem is still the nonce handling (for trusted calls).

In `tee-worker/cli/demo_shielding_unshielding.sh` there're three back-to-back indirect trusted calls:
- shield-funds (...wait for 30s)
- transfer
- unshield-funds

`shield-funds` uses `Teerex::shield_funds` directly, therefore the trusedCall is constructed upon block import and signed with the enclave signer.
trustedCalls of both `transfer` and `unshield-funds` are constructed from cli handling, `get_layer_two_nonce` is used to get the nonce. It dispatches a synchronous rpc call to retrieve the `System::nonce` state. Here is where the race condition kicks in, at the time of call, if the previous trustedCall is still pending, the nonce won't be incremented. It's easy to happen between `transfer` and `unshield-funds`, as there's no waiting time in between.

As a result, the `transfer` will be executed successfully (although the subsequent getter shows the balance is unchanged, but that's because the STF is not yet executed), while `unshield-funds` will fail -- you will see a `InvalidNonce` in worker logs.

The same problem happened upstream too: https://github.com/integritee-network/worker/actions/runs/3901167338

Basically, there're two solutions to this problem:
1. wait until the desired result (without timeout) in the testing script before continuing
2. similar to the enclave signer signing, consider pending tx in `get_layer_two_nonce` too - the precondition is the previous trustedCall was submitted to the top pool already.

However, these are still short-term solutions, imagine some trustedCall fails to execute, then the pre-computed nonce is guaranteed to differ from the actual nonce. It seems inevitable as long as the extrinsic/TrustedCall is constructed offline. We should consider supporting nonce = -1 (automatic nonce handling)

**Comment by Kailai-Wang** [comment]
Or maybe we need both solutions, plus the third: the nonce should be incremented at the beginning of trustedCall execution, irrelevant to the execution status, as the trustedCall is considered to be in the pool already.

**Comment by Kailai-Wang** [comment]
Maybe 3 is not as easy as I thought, the tests didn't go right -- maybe there's special handling for failed transaction. It seems it won't be included in the block at all? Was it removed somewhere?
